docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,196 @@
+
+### User Stories
+
+--------------------------------------------
+| ID | Story                               |
+|----|-------------------------------------|
+| U1 | Cluster Autoscaler                  |
+| U2 | MachinePool Machine Remediation     |
+| U3 | MachinePool Machine Rolling Updates |
+--------------------------------------------
+
+#### Story U1
+
+A cluster admin is tasked with using Cluster Autoscaler to scale a MachinePool as required by machine load. The MachinePool should provide a mechanism for Cluster Autoscaler to increase replica count, and provide a mechanism for Cluster Autoscaler to reduce replica count by selecting an optimal machine for deletion.
+
+#### Story U2
+
+A cluster admin would like to configure CAPI to remediate unhealthy MachinePool Machines to ensure a healthy fleet of Machines.
+
+#### Story U3
+
+A cluster admin updates a MachinePool to a newer Kubernetes version and would like to configure the strategy for that deployment so that the MachinePool will progressively roll out the new version of the machines. They would like this operation to cordon and drain each node to minimize workload disruptions.
+
+### Requirements
+
+--------------------------------------
+| ID | Requirement | Related Stories |
+|----|-------------|-----------------|
+| R1 | The MachinePool controller MUST create Machines representing the providerIDs<br>of the Machines in the MachinePool and enable cascading delete of infrastructure<br>machine resources upon delete of CAPI Machine. | U1 |
+| R2 | The machine health check controller MUST be able to select machines by label which belong to a MachinePool and remediate. | U2 |
+| R3 | The MachinePool API MUST provide an optional deployment strategy using the same type as MachineDeployment.Spec.Strategy. | U3 |
+| R4 | The Machine Controller MUST handle node drain for Machine Pool Machines with the same behavior as MachineDeployment Machines. | U1, U3 |
+
+### Implementation Details/Notes/Constraints
+
+- As an alternative, the `infrastructureRefList` field could be attached to the provider-specific MachinePool resource. Then the CAPI MachinePool controller could check through the provider-specific MachinePool resource through its infraRef to determine if it has an `infrastructureRefList` and use it if so. This would be very similar to the current design, except it would avoid changing the MachinePool spec in CAPI.
+- The `infrastructureRefList` might be more appropriate on MachinePool.Status instead of MachinePool.Spec.
+- It is possible that the `infrastructureRefList` could become large, if there are many machines in the MachinePool. It's a small struct, but carrying a long list around isn't optimal. One alternative suggested is to use a Kubernetes Selector similarly to MachineDeployment, but it's unclear how it would be used since the code in question is creating Machines, not finding existing ones.
```

Comment:
Yup if possible we should avoid having the list in the MachinePool.

Related Q: If it might be already a problem if we have a list of thousands of Machines in a MachinePool resource, wouldn't we have an even worse problem with thousands of Machine resources?
(I'm aware that avoiding the list is a best-effort performance optimization, while having the Machine resources is the desired outcome, but I still wonder what happens in that case)

But overall it sounds like the best we can do is try to avoid the list.

Reply:
If we can avoid the infrastructureRefList, we could also avoid the providerIDList at least in theory, because it already may suffer from a similar scaling issue. Removing an existing field would be harder to push forward given existing implementations, but we could plan for it perhaps outside of this proposal.